src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs

```diff
@@ -90,99 +72,147 @@ internal sealed class TupleTypeSymbol : NamedTypeSymbol
 
             NamedTypeSymbol underlyingType = GetTupleUnderlyingType(elementTypes, syntax, compilation, diagnostics);
 
-            return new TupleTypeSymbol(underlyingType, elementNames, compilation.Assembly);
+            return Create(locationOpt, underlyingType, elementLocations, elementNames);
         }
 
-        /// <summary>
-        /// Copy the original tuple, but modify it to use the new underlying type.
-        /// </summary>
-        private TupleTypeSymbol(TupleTypeSymbol originalTuple,
-                                NamedTypeSymbol newUnderlyingType)
+        public static TupleTypeSymbol Create(Location locationOpt, NamedTypeSymbol tupleCompatibleType, ImmutableArray<Location> elementLocations, ImmutableArray<string> elementNames)
         {
-            _underlyingType = newUnderlyingType;
-            _hasFriendlyNames = originalTuple._hasFriendlyNames;
+            return Create(locationOpt == null ? ImmutableArray<Location>.Empty : ImmutableArray.Create(locationOpt),
+                          tupleCompatibleType, elementLocations, elementNames);
+        }
 
-            var fieldsBuilder = ArrayBuilder<TupleFieldSymbol>.GetInstance(originalTuple._fields.Length);
+        public static TupleTypeSymbol Create(ImmutableArray<Location> locations, NamedTypeSymbol tupleCompatibleType, ImmutableArray<Location> elementLocations, ImmutableArray<string> elementNames)
+        {
+            Debug.Assert(tupleCompatibleType.IsTupleCompatible() && !tupleCompatibleType.IsTupleType);
 
-            int fieldIndex = 0;
-            int remainder = 0;
-            NamedTypeSymbol currentLinkType = newUnderlyingType;
+            ImmutableArray<TypeSymbol> elementTypes;
 
-            while (true)
+            if (tupleCompatibleType.Arity == RestPosition)
+            {
+                // Ensure all Rest extensions are tuples
+                tupleCompatibleType = EnsureRestExtensionsAreTuples(tupleCompatibleType);
+
+                var extensionTupleElementTypes = tupleCompatibleType.TypeArgumentsNoUseSiteDiagnostics[RestPosition - 1].TupleElementTypes;
+                var typesBuilder = ArrayBuilder<TypeSymbol>.GetInstance(RestPosition - 1 + extensionTupleElementTypes.Length);
+                typesBuilder.AddRange(tupleCompatibleType.TypeArgumentsNoUseSiteDiagnostics, RestPosition - 1);
+                typesBuilder.AddRange(extensionTupleElementTypes);
+                elementTypes = typesBuilder.ToImmutableAndFree();
+            }
+            else
+            {
+                elementTypes = tupleCompatibleType.TypeArgumentsNoUseSiteDiagnostics;
+            }
+
+            return new TupleTypeSymbol(locations, tupleCompatibleType, elementLocations, elementNames, elementTypes);
+        }
+
+        private static NamedTypeSymbol EnsureRestExtensionsAreTuples(NamedTypeSymbol tupleCompatibleType)
```

Comment:
Sure, I'll try to come up with something. I encourage you to provide an example of what you would put there to make sure your specific concern, if any, is addressed.

---
In reply to: [61807571](https://github.com/dotnet/roslyn/pull/11006#discussion_r61807571) [](ancestors = 61807571)